### PR TITLE
Raise on endpoint error handling, enable it by default in UCX 1.11

### PR DESCRIPTION
**Pull request by pentschev**

### CODE REVIEW

**Comment by quasiben** [comment]
This PR has all tests passing.  Are you still working on it ?

**Comment by pentschev** [comment]
This is still not fully working, but I'm trying to get remaining issues rounded up. However, it would still be great to have your opinion here @madsbk , in particular with the change I had to make on creating endpoints.

In latest changes I moved `UCXWorker.ep_create*` to `UCXEndpoint.ep_create*` class methods, which is required so I can now pass `UCXEndpoint`'s `self` to the error handling callback function so it can cancel inflight messages, which otherwise makes it impossible to raise when an error occurs during `UCXEndpoint.tag_send_nb`/`UCXEndpoint.tag_recv_nb`.

Do you see any problems in doing the change above?

**Comment by madsbk** [comment]
> In latest changes I moved `UCXWorker.ep_create*` to `UCXEndpoint.ep_create*` class methods, which is required so I can now pass `UCXEndpoint`'s `self` to the error handling callback function so it can cancel inflight messages, which otherwise makes it impossible to raise when an error occurs during `UCXEndpoint.tag_send_nb`/`UCXEndpoint.tag_recv_nb`.
> 
> Do you see any problems in doing the change above?

I cannot see why this is necessary? but I am fine with having the endpoint creation methods as class methods if we remove the `ep_` prefix :)

**Comment by pentschev** [comment]
> I cannot see why this is necessary?

It is necessary because I now [pass the `UCXEndpoint`'s `self` attribute to the error callback](https://github.com/rapidsai/ucx-py/pull/693/files#diff-fcd6a3e983ecdd8df6934febc5776f1122186708a3dbb76bee29acd33f2b86caR662). This is later used in the callback to [assign the current error status of an endpoint](https://github.com/rapidsai/ucx-py/pull/693/files#diff-fcd6a3e983ecdd8df6934febc5776f1122186708a3dbb76bee29acd33f2b86caR336-R337) and [cancel inflight messages](https://github.com/rapidsai/ucx-py/pull/693/files#diff-fcd6a3e983ecdd8df6934febc5776f1122186708a3dbb76bee29acd33f2b86caR349).

> but I am fine with having the endpoint creation methods as class methods if we remove the `ep_` prefix :)

Good point, this is done now.

**Comment by pentschev** [comment]
I believe this is finally good now. On local testing this now seems to work properly, including Dask with endpoint error handling with InfiniBand -- one of the main motivators for this change, avoiding nasty hangs when the other endpoint has closed. Note that endpoint error handling is only active by default in UCX 1.11 (current master) and above, because CUDA IPC error handling is only introduced in https://github.com/openucx/ucx/pull/6352 , and before that CUDA IPC would be quietly disabled if error handling was active.

@madsbk @quasiben would you be able to take a look?

File: ucp/_libs/ucx_endpoint.pyx [hunk 1]
```diff
@@ -36,7 +86,12 @@ def _ucx_endpoint_finalizer(uintptr_t handle_as_int, worker, set inflight_msgs):
     # Close the endpoint
     # TODO: Support UCP_EP_CLOSE_MODE_FORCE
```
Comment (madsbk) [inline ucp/_libs/ucx_endpoint.pyx:87]:
```suggestion
```

File: ucp/_libs/ucx_endpoint.pyx [hunk 2]
```diff
@@ -50,39 +105,170 @@ cdef class UCXEndpoint(UCXObject):
...
+        status, _ = self._get_status_and_str()
+
+        return status == UCS_OK
+
+    def raise_on_error(self):
+        if not self._endpoint_error_handling:
+            return
+
+        status, status_str = self._get_status_and_str()
+
```
Comment (madsbk) [inline ucp/_libs/ucx_endpoint.pyx:254]:
```suggestion
```

**Review by madsbk — CHANGES_REQUESTED** [review]
LGTM @pentschev, nice work!
I only have some few suggestions

File: ucp/_libs/ucx_endpoint.pyx (hunk 1 above)
Comment (pentschev) [reply ucp/_libs/ucx_endpoint.pyx]: Done.

File: ucp/_libs/ucx_endpoint.pyx [hunk 3]
```diff
@@ -50,39 +105,170 @@ cdef class UCXEndpoint(UCXObject):
...
+        else:
+            params.err_mode = UCP_ERR_HANDLING_MODE_PEER
+        params.err_handler.cb = err_cb
+        params.err_handler.arg = <void *>self
+
+        cdef ucp_ep_h ucp_ep
+        cdef ucs_status_t status = ucp_ep_create(worker._handle, params, &ucp_ep)
+        assert_ucs_status(status)
+
+        free(<void *>params)
```
[the remark this answers is not in the document]
Comment (pentschev) [reply ucp/_libs/ucx_endpoint.pyx:144]: I think you're mixing up the objects. Notice here `params` is only used within the constructor itself when passed to `ucp_ep_create` which is synchronous, and there's no `c_util_sockaddr_free` in the constructor, therefore freeing it just after `ucp_ep_create` should be fine.

File: ucp/_libs/ucx_endpoint.pyx [hunk 4]
```diff
@@ -50,39 +105,170 @@ cdef class UCXEndpoint(UCXObject):
...
+            UCP_EP_PARAM_FIELD_ERR_HANDLER
+        )
+        params.flags = UCP_EP_PARAMS_FLAGS_CLIENT_SERVER
+        if c_util_set_sockaddr(&params.sockaddr, ip_address.encode(), port):
+            raise MemoryError("Failed allocation of sockaddr")
+
+        try:
+            ret = cls(worker, <uintptr_t>params, endpoint_error_handling)
+        finally:
+            c_util_sockaddr_free(&params.sockaddr)
```
[the remark this answers is not in the document]
Comment (pentschev) [reply ucp/_libs/ucx_endpoint.pyx:186]: Nice catch! Done.

File: ucp/_libs/ucx_endpoint.pyx [hunk 5]
```diff
@@ -50,39 +105,170 @@ cdef class UCXEndpoint(UCXObject):
...
+        params.field_mask = (
+            UCP_EP_PARAM_FIELD_FLAGS |
+            UCP_EP_PARAM_FIELD_CONN_REQUEST |
+            UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE |
+            UCP_EP_PARAM_FIELD_ERR_HANDLER
+        )
+        params.flags = UCP_EP_PARAMS_FLAGS_NO_LOOPBACK
+        params.conn_request = <ucp_conn_request_h> conn_request
+
+        return cls(worker, <uintptr_t>params, endpoint_error_handling)
```
[the remark this answers is not in the document]
Comment (pentschev) [reply ucp/_libs/ucx_endpoint.pyx]: Another nice catch! Done.

File: ucp/_libs/ucx_endpoint.pyx (hunk 2 above)
Comment (pentschev) [reply ucp/_libs/ucx_endpoint.pyx]: Done.

**Comment by pentschev** [comment]
Thanks for the review @madsbk , I think I addressed all your comments. Would you mind taking another look?

File: ucp/_libs/ucx_endpoint.pyx (hunk 3 above)
[the remark this answers is not in the document]
Comment (pentschev) [reply ucp/_libs/ucx_endpoint.pyx]: You were right all along. I missed the fact that `params` in the constructor _is_ the `params` object passed from the `classmethod` implementations. Fixed now.

**Review by madsbk — APPROVED** [review]
Looks good @pentschev, thanks for working on this!

**Comment by pentschev** [comment]
Thanks @madsbk !

**Pull request merged**